### ISSUE TITLE
laplace_transform of DiracDelta yields wrong result

### ISSUE THREAD

**Issue by josephcslater** [sympy/sympy#10840]
`laplace_transform(DiracDelta(t),t,s)`
should yield 1, but instead yields 1/2 . 

version 0.7.6.1 of sympy.

Example at
https://github.com/josephcslater/iPythonExamples/blob/master/Laplace_Transforms_Solution.ipynb

**Comment by jksuom** [comment]
Laplace transform is generally defined only for locally integrable functions satisfying certain growth conditions. It is possible to say that the formal integral (from ``-oo`` to ``oo``) denoting the two-sided Laplace transform of Dirac delta is equal to 1. However, the two one-sided transforms (from ``-oo`` to ``0`` and from ``0`` to ``oo``) are undefined since Dirac delta is not a function on the intervals. But if one would want to assign some value to them, 1/2 for each could be a reasonable choice.

**Comment by josephcslater** [comment]
A couple of points: 
a) The behavior is inconsistent with standard definitions of the Laplace Transform of a Dirac Delta. The result of 1 is fundamental to control theory and communication theory.
b) The result returned is inconsistent with Mathematica, Matlab (by extension Maple). I have only so many places to test it, though. 
c) The argument you have is that "half of the delta" is before zero. I see the point you make, but it is contrary to established convention and usage. Capturing half of it in the domain is quirky. 
b) The Dirac delta is the result of the time derivative of the Heaviside Function. The Laplace Transform of the Heaviside function per SymPy is 1/s. The Laplace Transform of the derivative of a function in the Laplace domain is s times the Laplace transform of the original function (minus initial conditions for a one-sided transform). 

Supporting references would be exhausting, but a very quick search gives:
http://ocw.mit.edu/courses/mathematics/18-03sc-differential-equations-fall-2011/unit-iii-fourier-series-and-laplace-transform/laplace-transform-basics/MIT18_03SCF11_s27_5text.pdf

http://www2.latech.edu/~schroder/slides/DE_slides/Laplace_transform_delta_fun.pdf 

https://www.khanacademy.org/math/differential-equations/laplace-transform/properties-of-laplace-transform/v/laplace-transform-of-the-dirac-delta-function

A result of 1 is the expected outcome of users, either based on convention of preceding software or those who use it in practice (I can find dozens of texts in control and signal processing theory with this result in their table). So, I appreciate the theoretical problem, but the practical solution is for the transform to be conducted inclusively of what occurs at zero (Dirac(t) exists neither prior nor after). Staking this position will certainly inhibit adoption in multiple fields. Users will  not care about the subtlety, which I believe is philosophically correct, but practically wrong.

**Comment by zanzibar7** [comment]
Currently, we yield the following, which you probably dislike even more. 
```
In [18]: laplace_transform(DiracDelta(t),t,s)
Out[18]: (-Heaviside(0) + 1, -oo, True)
```
We had a related conversation about this a little while back regarding how to deal with `Heaviside(0)` (see issue #9907), and even how to handle `1**oo`.  Specifically, in my research, evaluating `Heaviside(0)` at all in my work yields in-correct answers without work-arounds because the `Heaviside` is representing the indeterminate slope of a line tangent to the `Max` function at `x=0`.

I like the idea of fixing this, but think we should try to constrain the changes to be specific to the context of Laplace transforms, if we can.  In Laplace transforms, the implicit time-directionality there is clear.  And we don't want to ever give the wrong answer in other contexts that we imagine are less important.  

So, to fix, I propose `LaplaceTransform`  treat `Heaviside(0) == 0` in appropriate situations.

**Comment by josephcslater** [comment]
That would certainly address my concern, and context-based solutions are a good compromise. It's unfortunate that the standard is already set. As a student, I as a bit perplexed by the standard, looking at the current SymPy response as what I would have expected. Unfortunately applied convention has indeed been so well established as to be entrenched. It's not feasible, reasonable, or prudent for anyone writing a general theory book to cover the, at print time, possibility of variations in this result. 

I haven't looked, but the Fourier Transform likely has the same issue with the same needed resolution.

**Comment by hanspi42** [comment]
The point with the one-sided Laplace transform is that it is implicitely defined with integration boundaries "0-" to infinity, not 0 to infinity. The "0-" means infinitesimally smaller than zero. Therefore, if you have a Dirac impulse at zero, then all of its area is counted and the result is 1, not 1/2.

The idea behind it is that if you to the laplace transform of sp.DiracDelta(t-tau) and then let tau go to zero, then there is no discontinuity and the result is always 1.

I really warmly suggest this behaviour because most people who actually use the Laplace transform in their professional lives would expext there to be no curious discontinuities when going from almost ideal cases to ideal cases. This is also why virtually any table of Laplace transforms you can find will give you 1 as the Laplace transform of delta(t). 

By the way, the half-half view would come into play when you define the Heaviside function as the integral across the dirac impulse. Then a definition of Heviside(t) as 0 for t<0, 1/2 for t=0, and 1 for t>0 would make a lot of sense.

Therefore, in the context of the Laplace transform, it would be reasonable if Heaviside(0) was supposed to be 1/2 instead of undefined. This might solve a whole bunch of other open issues. It would disagree with some definitions of the Heaviside function, but agree with the one used by most practitioners.

Speaking of which, Heaviside(0)=1/2 would be advantageous anyway, because then sp.heaviside(t) is the same as (sp.sign(t)+1)/2.

**Comment by jksuom** [comment]
> integration boundaries "0-" to infinity, not 0 to infinity. 

This could probably be implemented by changing this line https://github.com/sympy/sympy/blob/ab864967e71c950a15771bb6c3723636026ba876/sympy/integrals/transforms.py#L976
to something like

    F = limit(integrate(exp(-s*t) * f, (t, u, oo)), u, 0, dir='-')

**Comment by hanspi42** [comment]
I tried the following:

```
def _laplace_transform(f, t, s_, simplify=True):
    """ The backend function for Laplace transforms. """
    from sympy import (re, Max, exp, pi, Min, periodic_argument as arg_,
                       arg, cos, Wild, symbols, polar_lift)
    s = Dummy('s')
    u = Dummy('u')
    F = limit(integrate(exp(-s*t) * f, (t, u, oo)), u, 0, dir='-')
```

I let the following run:

```
import sympy as sp 
t,s = sp.symbols('t,s')
print(sp.laplace_transform(sp.DiracDelta(t),t,s))
```

Without the change, sympy 1.6.2, the output is `(1 - Heaviside(0), -oo, True)`, with the change it is `(1, -oo, True)`

This is great. I will think about more test cases.

**Comment by hanspi42** [comment]
Inverse Laplace is already OK in sympy 1.6.2, which is curious because it is inconsistent ...

```
import sympy as sp 
t,s = sp.symbols('t,s')
print(sp.inverse_laplace_transform(1,s,t))
```

gives `DiracDelta(t)`

**Comment by zanzibar7** [comment]
Using the lower limit is one of the ideas I proposed --  I think it's good.  But there might be an efficency hit from introducing another limit, and this breaks some current results.
```
s,u = symbols('s,u', real=True)
f = exp(-1/t)
integrate(exp(-s*t) * f, (t, 0, oo))
limit(integrate(exp(-s*t) * f, (t, u, oo)), u, 0, dir='-')
```
(Not that this is a particularly useful result, but who are we to judge)

**Comment by hanspi42** [comment]
This means you want to calculate the Laplace transform of `exp(-1/t)`?  

```
import sympy as sp 
t,s = sp.symbols('t,s')
print(sp.laplace_transform(sp.exp(-1/t),t,s))
```

gives `(meijerg(((), ()), ((-1, 0), ()), s), 0, True)` in SymPy 1.6.2. If I try it in my modified sympy install, I just get a `NotImplementedError: Don't know how to calculate the mrv of 'Integral(exp(-_s*t - 1/t), (t, -1/_p, oo))'`. Does that mean the Gruntz algorithm is only used here when `dir='-'` is given?

Does this mean that the Gruntz algorithm is **not** used here when `dir='-'` is not given?

**Comment by zanzibar7** [comment]
Yeah.  exp(-1/t) has a well-defined laplace transform when we evaluate the integral over [0,oo) or (0,oo).  But exp(-1/t) has an essential singularity at t=0, so the indefinite integral isn't defined at all for negative lower bounds.  As we try to evaluate it's Laplace transform with the lower limit approaching 0 from below, the limit is no longer defined, atleast in a classical sense.  It doesn't matter what algorithm is being used to evaluate the integral.

Something like an if-statement that only tries to take this lower-limit when there's a Heaviside in the laplace-transform's result would avoid the problem.  But it's also special-casing.

Another option would be to rewrite all heavisides so the convention H(0)=1 is honored.

Or we could modify DiracDelta to all a more consistent definition for the Laplace transform.  While it's common to assume the DiracDelta is symmetric, that isn't strictly necessary.  If the DiracDelta can be given an option for sided-ness, then that would be inherited by Heaviside,
and the evaluation would work automatically, without having to rewrite anything in laplace_transform.

**Comment by hanspi42** [comment]
"Another option would be to rewrite all heavisides so the convention H(0)=1 is honored."

That would be incorrect. I do not even want to get into any general discussions about the Heaviside step function for all applications, but in the particular context of integral transforms and the relationship between multiplication in one domain to convolution in the other domain, the correct choice is H(0)=1/2.

Proof by example: when you integrate 2 cos(t) sinc(t) from -oo to +oo, you can solve it by taking take the Fourier transform of 2 cos(t) sinc(t) and setting omega=0. This example is constructed deliberately to frequency-shift the tipping points of two Heaviside functions to omega=0, and the result will be 2*pi*H(0). The correct result calculated through other integration methods is pi, therefore H(0)=1/2 is necessary for consistency of Fourier transform theory (and with it also Laplace and I guess Mellin).

More details: http://schmid-werren.ch/hanspeter/publications/2014elemath.pdf , page 17.

Question: would it be an option to add a particular version of the Heaviside function to sympy that is compatible with the convolution theorem of integral transforms? It would be a Heaviside with H(0)=1/2 and we could call it StepFunction or similar. Then a Fourier transform of sinc(t) would be a sum ofg two frequency-shifted StepFunction().

If this is possible, I think we could much more easily take everything else from there without breaking existing code.

**Comment by zanzibar7** [comment]
I didn't mean for all transforms or applications -- just the Laplace transform.  It seems like a competing-limits issue.  I agree
that H(0)=1/2 is natural in many contexts, but like Joseph suggests up-top, H(0)=0 would be most consistent with convention in his control literature (and hence, your limit suggestion).

The `sign(x)` function is already implemented, and since sign(0)=0 in it's definition, it wouldn't take much to treat Heaviside() in terms of sign() and recover the H(0)=1/2 behavior.  Would that be good enough?

The other annoying issue right now is that `Integral(Heaviside(x-3),(x,0,9)).doit()` has an answer expressed in terms of `meijerg` rather than just `6`.

**Comment by oscarbenjamin** [comment]
> would it be an option to add a particular version of the Heaviside function

There already is if you use the second argument to Heaviside to specify the value of `H(0)`:
```julia
In [9]: Heaviside(x, S.Half)
Out[9]: θ(x, 1/2)

In [10]: Heaviside(x, S.Half).subs(x, 0)
Out[10]: 1/2
```

**Comment by hanspi42** [comment]
> There already is if you use the second argument to Heaviside to specify the value of `H(0)`:

This one does not work properly with the integral transaforms, though.

```
import sympy as sp 
t,omega = sp.symbols('t,omega')
print(sp.fourier_transform(2*sp.sinc(2*sp.pi*t),t,omega)) 
F = sp.Heaviside(omega+1,sp.S.Half)-sp.Heaviside(omega-1,sp.S.Half)
print(sp.inverse_fourier_transform(F,omega,t)) 
F = sp.Heaviside(omega+1)-sp.Heaviside(omega-1)
print(sp.inverse_fourier_transform(F,omega,t)) 
```

gives

```
Piecewise((0, Abs(omega**2) > 1), (1, True))
-InverseFourierTransform(Heaviside(omega - 1, 1/2), omega, t) + InverseFourierTransform(Heaviside(omega + 1, 1/2), omega, t)
sin(2*pi*t)/(pi*t)
```

You also see that the first result is also slightly off, it would fail for tge 2*cos(t)*sinc(t) case I described above.

**Comment by zanzibar7** [comment]
That's another part of what needs fixing.  So, can you write-up a set of unit-tests for `test_transforms.py` that will check for the behavior you want?

**Comment by hanspi42** [comment]
Yes, I will, but I have two "how-to" questions, and then I'm operational:

1. How to start the tests?  I went to `site-packages\sympy\integrals` and did `python -m unittest`, but that will only run the 14 ruby tests. Could you tell me how I can run just the `test_transforms.py`?
2. How shall I deliver? Fork sympy and tell you where my proposal is? Just tell here or pull request, and if yes, to what branch?

**Comment by jksuom** [comment]
> how I can run just the test_transforms.py?

One possibility is `bin/test transforms` though `bin/test` may have to be replaced by something else in Windows.

**Comment by sidhu1012** [comment]
>  Could you tell me how I can run just the `test_transforms.py`?

You could do pytest filepath
Example:-
pytest sympy/../test_transforms.py

**Comment by oscarbenjamin** [comment]
I use pytest. But you can do this from within python with
```julia
In [1]: import sympy

In [2]: sympy.test('test_transforms.py')
```

**Comment by hanspi42** [comment]
Works well, thanks, but is it interntional that about half of the tests fail?

`11 failed, 12 passed, 1 skipped, 2 warnings in **51.85s`

**Comment by sidhu1012** [comment]
Only xfail are meant to fail.
Otherwise something is wrong

**Comment by hanspi42** [comment]
I first tried it on unmodified code ;)  

But thanks, it is just pytest that makes the tests fail, `sympy.test('test_transforms.py')` works.

So now I will start working on the test cases.

**Comment by oscarbenjamin** [comment]
I don't get any failures from pytest:
```console
$ pytest sympy/integrals/tests/test_transforms.py 
============================================================= test session starts =============================================================
platform darwin -- Python 3.8.1, pytest-5.4.3, py-1.8.1, pluggy-0.13.1
architecture: 64-bit
cache:        yes
ground types: gmpy 2.0.8

rootdir: /Users/enojb/current/sympy/sympy, inifile: pytest.ini
plugins: doctestplus-0.7.0, instafail-0.4.2, xonsh-0.9.13.dev1, cov-2.10.0, profiling-1.7.0, forked-1.1.3, xdist-1.32.0
collected 24 items                                                                                                                            

sympy/integrals/tests/test_transforms.py ...s....................                                                                       [100%]

================================================== 23 passed, 1 skipped in 242.67s (0:04:02) ==================================================
```
Note that a key difference between pytest and the other ways of running the tests is that it also runs slow tests by default so I usually run it with `-m 'not slow'` to skip those. Also if you have `pytest-xdist` installed then you can run with e.g. 4 cores by passing `-n4` so:
```console
$ pytest -m 'not slow' -n4 sympy/integrals/tests/test_transforms.py 
============================================================= test session starts =============================================================
platform darwin -- Python 3.8.1, pytest-5.4.3, py-1.8.1, pluggy-0.13.1
architecture: 64-bit
cache:        yes
ground types: gmpy 2.0.8

rootdir: /Users/enojb/current/sympy/sympy, inifile: pytest.ini
plugins: doctestplus-0.7.0, instafail-0.4.2, xonsh-0.9.13.dev1, cov-2.10.0, profiling-1.7.0, forked-1.1.3, xdist-1.32.0
gw0 [18] / gw1 [18] / gw2 [18] / gw3 [18]
..................                                                                                                                      [100%]
============================================================= 18 passed in 59.88s =============================================================
```

**Comment by hanspi42** [comment]
Thanks! I have no clue what it was. I have made a minimal virtual environment, and now it is fine. Thanks for supporting me, I am a newby with tests.

**Comment by jksuom** [comment]
As @oscarbenjamin [points out]( https://github.com/sympy/sympy/issues/10840#issuecomment-716032771), it is possible to set `H0` to any desired value. I would consider making use of that in `synpy/integrals/deltafunctions.py` by setting the value `0` on the two lines where `Heaviside` is returned:

    return Heaviside(f.args[0], 0)
    return r*Heaviside(x - point, 0)

**Comment by hanspi42** [comment]
> 
> 
> As @oscarbenjamin [points out](https://github.com/sympy/sympy/issues/10840#issuecomment-716032771), it is possible to set `H0` to any desired value. I would consider making use of that in `synpy/integrals/deltafunctions.py` by setting the value `0` on the two lines where `Heaviside` is returned:
> 
> ```
> return Heaviside(f.args[0], 0)
> return r*Heaviside(x - point, 0)
> ```

While it would possibly solve some problems, it would be mathemathically inconsistent. See above. It should be `S.Half`.
[https://github.com/sympy/sympy/issues/10840#issuecomment-715907908](url)

If you do 0, you will break what I described there and cause problems once sympy evolves enough to attack more complicated transforms.

It would be much better to choose `S.Half` and do your `dir='-'` solution above in a way that does not cause a fall-back to meijerg all the time.

**Comment by hanspi42** [comment]
I believe this issue is outdated and can be closed; the relevant discussion continues in #21372

**Comment by hanspi42** [comment]
> 
> 
> > integration boundaries "0-" to infinity, not 0 to infinity.
> 
> This could probably be implemented by changing this line
> 
> https://github.com/sympy/sympy/blob/ab864967e71c950a15771bb6c3723636026ba876/sympy/integrals/transforms.py#L976
> 
> to something like
> 
> ```
> F = limit(integrate(exp(-s*t) * f, (t, u, oo)), u, 0, dir='-')
> ```

I tested it with the changes made in https://github.com/sympy/sympy/pull/21452; this will work!  Maybe we now have a clear path to solving the large number of problems caused by the delta distribution.

```
from sympy import DiracDelta, symbols, integrate, oo, limit
t,tau = symbols('t,tau',real=True)
r = integrate(DiracDelta(t),(t,0,oo))
s = limit( integrate(DiracDelta(t),(t,tau,oo)), tau, 0, dir='-')
print(r)
print(s)
```
